src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -244,66 +248,123 @@ public String getUrl() throws IOException {
      *
      * This method provides such an opportunity for subtypes.
      */
-    protected void setActions(List<Action> actions) {
-        this.actions = new CopyOnWriteArrayList<Action>(actions);
+    protected synchronized void setActions(List<Action> actions) {
+            this.actions = new CopyOnWriteArrayList<Action>(actions);
     }
 
-/*
-    We can't use Actionable#actions to store actions because they aren't transient,
-    and we need to store actions elsewhere because this is the only mutable pat of FlowNode.
+    /**
+     * Return the first nontransient {@link Action} on the FlowNode, without consulting {@link jenkins.model.TransientActionFactory}s
+     * <p> This is not restricted to just Actions implementing {@link PersistentAction} but usually they should.
+     * Used here because it is much faster than base {@link #getAction(Class)} method.
+     * @param type Class of action
+     * @param <T>  Action type
+     * @return First nontransient action or null if not found.
+     */
+    @CheckForNull
+    @Restricted(NoExternalUse.class)  // Limit use to workflow-api packages until we have a case where we need the performance badly.
+    public final <T extends Action> T getPersistentAction(@Nonnull Class<T> type) {
```

Comment:
@jglick  Results in "type is not within its bound" compile error for the main getAction method invoking this.   Also not necessary, see javadocs: 

> This is not restricted to just Actions implementing {@link PersistentAction} but usually they should.